Timezone migration

dannyroberts opened the pull request:
opening for tests

dannyroberts committed:
make data processing halt mid timezone migration

and add helpers for getting and setting timezone progress

dannyroberts committed:
add timezone migration test

dannyroberts committed:
add test for pausing submissions during tz migration

dannyroberts committed:
tzmigrationtest (WIP)

dannyroberts committed:
tzmigration: save proposed changes to sqlite

dannyroberts committed:
tzmigration: verify sqlite row counts against couch WIP

dannyroberts committed:
fix tests after rebase

dannyroberts committed:
add spanning to tzmigrationtest

dannyroberts committed:
clean up tzmigrationtest mgmt cmd

and pull out functionality into callable chunks

dannyroberts committed:
have prepare_planning_db return something

dannyroberts committed:
fix types in planning db

dannyroberts committed:
fix confusing comprehension

dannyroberts committed:
extend tzmigration planner to reconstruct case

and save propsed new case_json

dannyroberts committed:
tweak case diffs

dannyroberts committed:
fix tests

dimagimon commented:
There are 5 errors:

* corehq/apps/tzmigration/management/commands/tzmigrationtest.py, line 51 - F841 local variable 'session' is assigned to but never used
* corehq/apps/tzmigration/management/commands/tzmigrationtest.py, line 52 - E702 multiple statements on one line (semicolon)
* corehq/apps/tzmigration/timezonemigration.py, line 13 - F401 'get_migration_status' imported but unused
* corehq/apps/tzmigration/timezonemigration.py, line 44 - E241 multiple spaces after ','
* corehq/ex-submodules/couchforms/util.py, line 27 - F401 'HttpException' imported but unused

dannyroberts committed:
respond to lintreview

dimagimon commented on corehq/apps/tzmigration/management/commands/tzmigrationtest.py:
```diff
@@ -0,0 +1,84 @@
+            self.show_diffs()
+        if options['play']:
+            from corehq.apps.tzmigration.planning import *
+            session = self.planning_db.Session()  # pylint: disable-msg=F841
```
F841 local variable 'session' is assigned to but never used

dimagimon commented on corehq/apps/tzmigration/management/commands/tzmigrationtest.py:
```diff
@@ -0,0 +1,84 @@
+        if options['play']:
+            from corehq.apps.tzmigration.planning import *
+            session = self.planning_db.Session()  # pylint: disable-msg=F841
+            import ipdb; ipdb.set_trace()  # pylint: disable-msg=E702
```
E702 multiple statements on one line (semicolon)

dimagimon commented on corehq/apps/tzmigration/management/commands/tzmigrationtest.py:
```diff
@@ -0,0 +1,84 @@
+        if options['play']:
+            from corehq.apps.tzmigration.planning import *
+            session = self.planning_db.Session()  # noqa
+            import ipdb; ipdb.set_trace()  # noqa
```
E702 multiple statements on one line (semicolon)

dannyroberts committed:
fix delete_planning_db

dannyroberts committed:
exclude CommCareCase-Deleted from tzmigration

dannyroberts commented:
@czue I think at this point this is worth reviewing and merging. There's two things I'd review

1. All the files that aren't in the tzmigration app—these are the most likely to actually affect the current workflow
2. prepare_planning_db corehq/apps/tzmigration/timezonemigration.py (and the corresponding models in planning.py), which is how we're going to be deciding what new json to replace current forms and cases with

dannyroberts commented:
Not going to run anything immediately, just thought this was a good stopping point getting this into master. Also, did this TDD and was really happy when I pushed a commit last week that made the test I'd written in early May pass.

czue commented on corehq/apps/tzmigration/api.py:
```diff
@@ -66,6 +73,19 @@ def phone_timezones_should_be_processed():
+_thread_local._force_phone_timezones_should_be_processed = False
+
+
+class force_phone_timezones_should_be_processed():
```
not blocking but as far as i know there's no reason to ever use old style classes. maybe you just like that it looks more like a function?

czue replied:
(they are also disappearing in python 3, not that that feels super relevant to us yet)

dannyroberts replied:
oops, no I probably just converted it from being a function and didn't notice. Fixed

czue commented on corehq/apps/tzmigration/management/commands/tzmigrationtest.py:
```diff
@@ -0,0 +1,89 @@
+        return True
+
+
+class Command(BaseCommand):
```
(didn't review this closely but looks fun/thorough)

czue commented on corehq/apps/tzmigration/planning.py:
```diff
@@ -0,0 +1,198 @@
+Base = declarative_base()
+
+
+class PlanningForm(Base):
```
(not blocking) curious how you decided to use sqlalchemy

dannyroberts replied:
Yeah I was initially just writing SQLite queries using the `sqlite3` module, but that was feeling pretty ridiculous.

czue commented on corehq/apps/tzmigration/planning.py:
```diff
@@ -0,0 +1,198 @@
+        self.db_filepath = db_filepath
+        self._connection = None
+        self.engine = create_engine(
+            'sqlite+pysqlite:///{}'.format(db_filepath), module=sqlite)
```
aha

dannyroberts replied:
Seems like you figured it out, but the point is to have this be a bit more quick and dirty, and flexible for meddling: if something gets messed up in the planning and I want to re-run it, I just delete the sqlite file and start over without having to think about how if effects any other domains I've run the migration on. I'm sure there are other ways, but this seemed simple and easy and silo'd and flexible

czue commented on corehq/apps/tzmigration/timezonemigration.py:
```diff
@@ -0,0 +1,190 @@
+def run_timezone_migration_for_domain(domain):
+    set_migration_started(domain)
+    _run_timezone_migration_for_domain(domain)
+    set_migration_complete(domain)
```
should we delete the planning db here? am I wrong to think that we're going to run out of space on the disk pretty quickly if we don't clean up?

dannyroberts replied:
I think at least until this is completely not manual, it's good to keep it around for a while. The db includes all the diffs, from which we can extract useful information. I think after the first few runs, I'll have a better idea of what exactly I want to be saving, and then we can dump that to a more permanent log file that we can refer to later in case someone complains about a change/inconsistency in their project data.

czue replied:
Sounds good
On Sep 3, 2015 4:22 PM, "Daniel Roberts" <notifications@github.com> wrote:

> In corehq/apps/tzmigration/timezonemigration.py
> <https://github.com/dimagi/commcare-hq/pull/8070#discussion_r38691985>:
>
> > +from corehq.apps.commtrack.processing import get_stock_actions
> > +from dimagi.utils.couch.database import iter_docs
> > +from corehq.apps.tzmigration import set_migration_started, \
> > +    set_migration_complete
> > +from corehq.apps.tzmigration.planning import PlanningDB
> > +from couchforms.dbaccessors import get_form_ids_by_type
> > +from couchforms import convert_xform_to_json
> > +from couchforms.models import XFormInstance
> > +from couchforms.util import adjust_datetimes, scrub_meta
> > +from corehq.apps.tzmigration import force_phone_timezones_should_be_processed
> > +
> > +
> > +def run_timezone_migration_for_domain(domain):
> > +    set_migration_started(domain)
> > +    _run_timezone_migration_for_domain(domain)
> > +    set_migration_complete(domain)
>
> I think at least until this is completely not manual, it's good to keep it
> around for a while. The db includes all the diffs, from which we can
> extract useful information. I think after the first few runs, I'll have a
> better idea of what exactly I want to be saving, and then we can dump that
> to a more permanent log file that we can refer to later in case someone
> complains about a change/inconsistency in their project data.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/8070/files#r38691985>.
>

czue commented:
wow this looks very thorough and paranoid - nice work! no real blockers that I saw and seemed like you had thought through things much more carefully than I was going to on review.

curious what the high level plan after this gets merged is. manually run it on a few domains. then what?

dannyroberts committed:
tidy up force_phone_timezones_should_be_processed

- explicit superclass
- use camelcase for class name, expose function returning an instance

dannyroberts commented:
I think it's run it on a few domains, and take note of what kinds of problems if any we run into, and then figure out what to do from there. Based on how things go I'll probably streamline it somewhat. The final process might be something like.

1. Click a button in the to start the process, and freeze the data (turning away any new submissions).
2. Get an overview of any significant data changes
3. Click either "Apply" or "Rollback"
  - Rollback just does nothing except start accepting data again
4. Apply will then iterate through the proposed changes and save them to the couch; as you've mentioned I'll need to introduce throttling here for this to be practical. I'm thinking I'll just throttle by bulk saving 500 forms/cases and then hitting a form/case view, and then repeating.

For domains that use stock, I still have some work to do. I'm thinking I'll do the intrahealth changes first if we commit to those since they're similar and relevant.

czue commented:
just waiting for tests but :+1: to merge when they pass

czue commented:
:boom: :bomb: :boom:
